java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -388,6 +389,57 @@ public ExchangeReaderWriter doExchange(FlightDescriptor descriptor, CallOption..
     }
   }
 
+  /**
+   * Wrapper class to wrap the iterator and handle auth failures and retry.
+   * @param <T> The type of iterator.
+   */
+  public static class WrappedFlightIterator<T> implements Iterator<T> {
```

Comment:
A general comment in the interests of code reuse; you may be better off implementing a generic StreamObserver<T> which can wrap any gRPC call. You'll have to change FlightClient to always use the async gRPC stub instead of the blocking stub, but then you can handle retries for any gRPC call without having to write special-purpose wrappers for each call.

Reply:
@lidavidm Thank you for your feedback

I have pushed a PS with the suggested changes.

I do currently have a blocking call [[link]](https://github.com/apache/arrow/pull/8780/commits/2fc2971979a1fa93346948d2da9cba91a8399c43#diff-5f196a1c597e5a8d9c3cd835dae4db594b6204cb2d681c494d8ed3906cf47478R47) on the API calls that return an Iterable, like listFlights and listActions, as we need to have the iterable populated before we return it to the FlightClient, if not then we are back to the same approach using a wrapped iterator.

While the changes latest PS `doClientCall ` only caters to API calls that return an iterable/iterator, the next step is to make them generic for calls like `getInfo, getSchema`
Before I go ahead and do that, I wanted to confirm if this is what you had in mind when you made the suggestions above, regarding a generic StreamObserver.

Other options that I had in mind were to pass the functional interfaces to pass the `ApiCallObserver` and have it retry the API call `onError`, in that case, we would not need to block `doClientCall` until the `ApiCallObserver` completes as the API call would be retried asynchronously when the Observer encounters an error.

Reply:
Sorry - let me clarify. What I mean is that gRPC offers 3 types of clients in Java: a blocking client, async client, and a "generic" client. This last client offers a unified interface for all types of calls: it accepts a StreamObserver which gets called with data from the server and it is given a StreamObserver to send data to the server. So if you were to rewrite all gRPC calls in terms of this last type of client, then you could implement a single StreamObserver wrapper that handles retries for all calls.

Not a requirement - just offering a suggestion to avoid having to implement lots of single-purpose retries. It would take a decent amount of refactoring.

Reply:
So I see you have switched to using the generic client; my comment is then more about the level of abstraction appropriate for implementing retry. In particular, keeping the logic inside a wrapper StreamObserver would, IMO, make it easier to preserve things like the streaming nature of calls without having to buffer results.